### PR TITLE
Floatingip

### PR DESCRIPTION
**Pull request by EarthmanT**
Modified this to reflect that connect/disconnect (previously attach/detach) will take place in the context of relationship lifecycle operations - so a mock relationship context in the tests, and source/target rest calls in the operations themselves.

### CODE REVIEW

File: ec2/floatingip.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+# Cloudify imports
+from cloudify import ctx
+from cloudify.exceptions import NonRecoverableError
+from cloudify.decorators import operation
+from ec2 import connection
+
+
+@operation
+def create(**kwargs):
+    """ Gets a floating IP from an Amazon Elastic IP.
```
Comment (nir0s) [inline ec2/floatingip.py]: Gets -> Provisions maybe?

File: ec2/floatingip.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    ec2_client = connection.EC2ConnectionClient().client()
+    ctx.logger.debug('Getting Elastic IP from Amazon EC2.')
```
Comment (nir0s) [inline ec2/floatingip.py]: Again... Getting is kinda weird.

Comment (nir0s) [inline ec2/floatingip.py]: Wait.. is "Get" the AWS name for this process?

File: ec2/floatingip.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    try:
+        address_object = ec2_client.allocate_address()
+    except (EC2ResponseError, BotoServerError) as e:
+        raise NonRecoverableError('(Node: {0}): Error. Failed to get '
+                                  'floating ip, returned: {1}.'
```
Comment (nir0s) [inline ec2/floatingip.py]: Um.. this is kinda relevant for the rest of the review.. .but didn't we decide that all resources are named according to their Cloud environment?

We shouldn't have "floating ip" anywhere.. only Elastic IP. or .. am I wrong?

File: ec2/floatingip.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                                  .format(ctx.instance.id, e))
+
+    ctx.logger.info('Elastic IP creation response: {0}'.format(address_object))
+    ctx.instance.runtime_properties['floatingip'] = address_object.public_ip
+
```
Comment (nir0s) [inline ec2/floatingip.py]: can you log something like:

```python
ctx.logger.info('Elastic IP Provisioned: {0}'.format(address_object.public_ip1))
```

?

I think this is true for all logging throughout the plugin.
We should state that we're:

```python
ctx.logger.debug('Getting Elastic IP from Amazon EC2.')
```

then, if successful:

```python
ctx.logger.info('Elastic IP Provisioned: {0}'.format(address_object.public_ip))
```

Note the logging level. These are all 
or if failed:

```python
ctx.logger.info('Failed to provision Elastic IP')
```

note the logging level.

File: ec2/floatingip.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+@operation
+def delete(**kwargs):
+    ec2_client = connection.EC2ConnectionClient().client()
```
Comment (nir0s) [inline ec2/floatingip.py]: missing docstring

File: ec2/floatingip.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    ctx.logger.info('Deleted Floating IP {0}.'.format(floatingip))
+
+
+@operation
+def connect(**kwargs):
+    """ Attaches a floating IP to a node.
+    """
+    ec2_client = connection.EC2ConnectionClient().client()
+    ctx.logger.debug('Attaching a Floating IP to EC2 Instance.')
```
Comment (nir0s) [inline ec2/floatingip.py]: This, for instance, should be changed if you want it to be a debug message.
It should state that it's trying to attach IP {0} to Instance {1}. Or.. maybe.. it's not needed..

File: ec2/floatingip.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                                  'floating ip, returned: {0}.'
+                                  .format(e))
+
+    ctx.logger.info('Attached Floating IP {0} to instance {1}.'.format(
+        floatingip, instance_id))
+
+
+@operation
+def disconnect(**kwargs):
+    ec2_client = connection.EC2ConnectionClient().client()
```
Comment (nir0s) [inline ec2/floatingip.py]: docstirng

File: ec2/tests/test_ec2.py [hunk 8]
```diff
@@ -24,11 +24,14 @@
...
 from cloudify.mocks import MockCloudifyContext
+from cloudify.mocks import MockContext
 from cloudify.exceptions import NonRecoverableError
 
 TEST_AMI_IMAGE_ID = 'ami-e214778a'
 TEST_INSTANCE_TYPE = 't1.micro'
+FQDN = '((?:[a-z][a-z\\.\\d\\-]+)\\.(?:[a-z][a-z\\-]+))(?![\\w\\.])'
 
 
 class TestPlugin(testtools.TestCase):
```
Comment (nir0s) [inline ec2/tests/test_ec2.py]: You should have different classes for testing different resources.

Maybe something like:

```python
class TestInstance(...)
   ...

class TestElasticip(...)
   ...

File: ec2/floatingip.py (hunk 1 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: I am now "provisioning"

File: ec2/floatingip.py (hunk 2 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: No. AWS "allocates"

File: ec2/tests/test_ec2.py (hunk 8 above)
Comment (EarthmanT) [inline ec2/tests/test_ec2.py]: Created test_ec2_*module*.py for each module and moved the related functions in there.
**I also did this for the utils module even though currently all utils relate to a particular module.

File: ec2/floatingip.py (hunk 7 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: added

File: ec2/floatingip.py (hunk 6 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: made it info

File: ec2/floatingip.py (hunk 5 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: added

File: ec2/floatingip.py (hunk 4 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: made the change to everything

File: ec2/floatingip.py (hunk 3 above)
Comment (EarthmanT) [inline ec2/floatingip.py]: made the change everywhere

File: ec2/elasticip.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+    try:
+        ec2_client.associate_address(instance_id=instance_id,
+                                     public_ip=elasticip)
+    except (EC2ResponseError, BotoServerError) as e:
+        raise NonRecoverableError('Error. Failed to '
+                                  'attach Elastic IP. Error: {0}.'
+                                  .format(e))
+
+    ctx.logger.info('Connected Elastic IP to instance.')
```
Comment (nir0s) [inline ec2/elasticip.py]: can you add the ip and the instance here?

File: ec2/elasticip.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    elasticip = ctx.instance.runtime_properties['elasticip']
+
+    try:
+        ec2_client.release_address(public_ip=elasticip)
+    except (EC2ResponseError, BotoServerError) as e:
+        raise NonRecoverableError('Error. Failed to '
+                                  'delete Elastic IP. Error: {0}.'
+                                  .format(e))
+
+    ctx.logger.info('Deleted Elastic IP {0}.')
```
Comment (nir0s) [inline ec2/elasticip.py]: following the convention .. this should be IP: {0} :)

File: ec2/elasticip.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    except (EC2ResponseError, BotoServerError) as e:
+        raise NonRecoverableError('Error. Failed to detach '
+                                  'Elastic IP, returned: {0}.'
+                                  .format(e))
+
+    ctx.logger.info('Disconnected Elastic IP from instance.')
+
+
+@operation
+def creation_validation(**kwargs):
```
Comment (nir0s) [inline ec2/elasticip.py]: did you verify with Idan that we still need this?

File: ec2/elasticip.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    elasticip = ctx.instance.runtime_properties['elasticip']
+
+    try:
+        ec2_client.get_all_addresses(elasticip)
+    except (EC2ResponseError, BotoServerError) as e:
+        raise NonRecoverableError('Error. Failed to validate '
+                                  'Elastic IP, returned: {0}.'
+                                  .format(e))
+
+    ctx.logger.info('Verified that Elastic IP was created.')
```
Comment (nir0s) [inline ec2/elasticip.py]: again... IP: {0}

File: ec2/tests/test_ec2_elasticip.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        return relationship_ctx
+
+    def test_create(self):
+        """
+        """
+        ctx = self.mock_ctx('test_create_address')
+
+        with mock_ec2():
+            elasticip.create(ctx=ctx)
+            self.assertTrue('elasticip' in ctx.instance.runtime_properties)
```
Comment (nir0s) [inline ec2/tests/test_ec2_elasticip.py]: should use assertIn rater than assertTrue(x in y). assertTrue only returns true or false while assertIn returns the exception.

File: ec2/utils.py [hunk 14]
```diff
@@ -118,5 +107,38 @@ def get_instance_from_id(instance_id, ctx):
...
-
     return instance
+
+
+def get_instance_variable(instance, variable):
+    """ given the related instance object
+        the given variable can be retrieved and returned
+    """
+    while instance.update() != 'running':
+        time.sleep(5)
```
Comment (nir0s) [inline ec2/utils.py]: requires a timeout

File: ec2/utils.py [hunk 15]
```diff
@@ -118,5 +107,38 @@ def get_instance_from_id(instance_id, ctx):
                                   .format(ctx.instance.id, e))
 
     instance = reservations[0].instances[0]
-
     return instance
+
+
+def get_instance_variable(instance, variable):
```
Comment (nir0s) [inline ec2/utils.py]: did you mean variable, or attribute?

File: ec2/elasticip.py (hunk 11 above)
Comment (EarthmanT) [inline ec2/elasticip.py]: He said for now yes.

File: ec2/utils.py (hunk 15 above)
Comment (EarthmanT) [inline ec2/utils.py]: changed to attribute

File: ec2/utils.py (hunk 14 above)
Comment (EarthmanT) [inline ec2/utils.py]: added configurable timeout_length

File: ec2/tests/test_ec2_elasticip.py (hunk 13 above)
Comment (EarthmanT) [inline ec2/tests/test_ec2_elasticip.py]: Switched to assertIn

File: ec2/elasticip.py (hunk 12 above)
Comment (EarthmanT) [inline ec2/elasticip.py]: done

File: ec2/elasticip.py (hunk 10 above)
Comment (EarthmanT) [inline ec2/elasticip.py]: added

File: ec2/elasticip.py (hunk 9 above)
Comment (EarthmanT) [inline ec2/elasticip.py]: switched to amazon names

**Pull request merged**